Allow .js preference files to set locked prefs with
 lockPref()

diff -u a/modules/libpref/prefapi.cpp b/modules/libpref/prefapi.cpp
--- a/modules/libpref/prefapi.cpp	2016-05-12 13:05:06.000000000 -0400
+++ b/modules/libpref/prefapi.cpp	2016-06-01 10:41:23.629568981 -0400
@@ -947,7 +947,8 @@
                          PrefValue   value,
                          PrefType    type,
                          bool        isDefault,
-                         bool        isStickyDefault)
+                         bool        isStickyDefault,
+                         bool        isLocked)
 
 {
     uint32_t flags = 0;
@@ -960,4 +961,6 @@
         flags |= kPrefForceSet;
     }
     pref_HashPref(pref, value, type, flags);
+    if (isLocked)
+        PREF_LockPref(pref, true);
 }
diff -u a/modules/libpref/prefapi.h b/modules/libpref/prefapi.h
--- a/modules/libpref/prefapi.h
+++ b/modules/libpref/prefapi.h
@@ -181,14 +181,15 @@ nsresult PREF_UnregisterCallback( const char* domain,
 /*
  * Used by nsPrefService as the callback function of the 'pref' parser
  */
 void PREF_ReaderCallback( void *closure,
                           const char *pref,
                           PrefValue   value,
                           PrefType    type,
                           bool        isDefault,
-                          bool        isStickyDefault);
+                          bool        isStickyDefault,
+                          bool        isLocked);
 
 #ifdef __cplusplus
 }
 #endif
 #endif
diff -u a/modules/libpref/prefread.cpp b/modules/libpref/prefread.cpp
--- a/modules/libpref/prefread.cpp	2016-05-12 13:05:06.000000000 -0400
+++ b/modules/libpref/prefread.cpp	2016-06-01 10:41:23.629568981 -0400
@@ -43,6 +43,7 @@
 #define BITS_PER_HEX_DIGIT      4
 
 static const char kUserPref[] = "user_pref";
+static const char kLockPref[] = "lockPref";
 static const char kPref[] = "pref";
 static const char kPrefSticky[] = "sticky_pref";
 static const char kTrue[] = "true";
@@ -131,7 +132,7 @@
         break;
     }
     (*ps->reader)(ps->closure, ps->lb, value, ps->vtype, ps->fdefault,
-                  ps->fstickydefault);
+                  ps->fstickydefault, ps->flock);
     return true;
 }
 
@@ -191,6 +192,7 @@
                 ps->vtype = PrefType::Invalid;
                 ps->fdefault = false;
                 ps->fstickydefault = false;
+                ps->flock = false;
             }
             switch (c) {
             case '/':       /* begin comment block or line? */
@@ -202,10 +204,13 @@
             case 'u':       /* indicating user_pref */
             case 's':       /* indicating sticky_pref */
             case 'p':       /* indicating pref */
+            case 'l':       /* indicating lockPref */
                 if (c == 'u') {
                   ps->smatch = kUserPref;
                 } else if (c == 's') {
                   ps->smatch = kPrefSticky;
+                } else if (c == 'l') {
+                  ps->smatch = kLockPref;
                 } else {
                   ps->smatch = kPref;
                 }
@@ -252,8 +257,10 @@
         /* name parsing */
         case PREF_PARSE_UNTIL_NAME:
             if (c == '\"' || c == '\'') {
-                ps->fdefault = (ps->smatch == kPref || ps->smatch == kPrefSticky);
+                ps->fdefault = (ps->smatch == kPref || ps->smatch == kPrefSticky
+                                || ps->smatch == kLockPref);
                 ps->fstickydefault = (ps->smatch == kPrefSticky);
+                ps->flock = (ps->smatch == kLockPref);
                 ps->quotechar = c;
                 ps->nextstate = PREF_PARSE_UNTIL_COMMA; /* return here when done */
                 state = PREF_PARSE_QUOTED_STRING;
diff -u a/modules/libpref/prefread.h b/modules/libpref/prefread.h
--- a/modules/libpref/prefread.h	2016-05-12 13:05:06.000000000 -0400
+++ b/modules/libpref/prefread.h	2016-06-01 10:41:23.629568981 -0400
@@ -34,7 +34,8 @@
                            PrefValue   val,
                            PrefType    type,
                            bool        defPref,
-                           bool        stickyPref);
+                           bool        stickyPref,
+                           bool        lockPref);
 
 /* structure fields are private */
 typedef struct PrefParseState {
@@ -56,6 +57,7 @@
     PrefType    vtype;      /* PREF_STRING,INT,BOOL          */
     bool        fdefault;   /* true if (default) pref        */
     bool        fstickydefault; /* true if (sticky) pref     */
+    bool        flock;      /* true if pref to be locked     */
 } PrefParseState;
 
 /**
